rdm/0004-installation-and-upgrade-workflow.md

```diff
@@ -0,0 +1,160 @@
+- As a developer, I want to be able to build a container image of my application.
+- As a developer, I want to be able to run the application and the services it depends on as containers.
+- As a developer, I want to be able to run the application in a development server, so that I can easily debug and profile my application.
+- As a developer, I want to be able to stop a running instance and its corresponding services without loosing the data that is alocated in them.
+- As a developer, I want to be able to clean up the environment and all the services, so that I can start from scratch.
+- As a developer, I want to be able to update my running instance, so that I can apply changes such as customizations.
+- As a developer, I want to be able to upgrade my running instance to a newer versions of Invenio RDM.
+
+## Detailed design
+
+### CLI and commands
+Invenio RDM installation and upgrade procedures can be carried out using a CLI (Command Line Interface). This CLI provides the following commands:
+
+- `init`
+- `build`
+- `setup`
+- `run`
+- `update`
+- `upgrade`
+- `destroy`
+
+**init**
+
+The `init` command would initialize the files and folders structure, resulting in a working instance. The initialization will be performed by prompting the user with a set of questions that allow to customize the system architecture (i.e. project name, Elasticsearch version, relational database system). This boostraping is implemented as a Cookiecutter module.
+
+The resulting file structure is as follows:
+
+- `invenio.cfg`: Contains all the configuration variables.
+- `pipenv`: Contains the python dependencies along with each corresponding version.
+- `static/`: Contains all the CSS and potential static content (e.g. images).
+- `templates/`: Contains the custom templates.
+- `Dockerfile.base`: Description of the process to install the Python dependencies.
+- `Dockerfile`: Description of the process to create the application image.
+- `docker-services.yml`: Contains the description of the containers for all needed services.
+- `docker-compose.yml`: Contains the service components architecture and orchestration for a semi-production deployment. It uses the services described in the `docker-services.yml` file.
+- `docker-compose.dev.yml`: Contains the service components architecture and orchestration for a development deployment. It uses the services described in the `docker-services.yml` file.
+
+**build**
+
+The `build` command should allow to build two types of environments: Development and semi-production.
+
+Nonetheless, independently from the environment type, the first step of the build process is to lock the application dependencies. Therefore, by using Pipenv, the CLI will generate a *lock* file that would contain the version and hash identifier of each dependency. For this use case the flags ``--lock`` or ``--no-lock`` are used, being the first the default one.
```

Comment:
Yes... :( The problem is also that the lock happens outside of docker build. Since the lock file has to be copied over to the container (so it can be tracked via VCS). So we must allow for ``--skip-lock``.